File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        ArrayList<Integer> node = new ArrayList<>();

        ListNode temp = head.next;
        ListNode prev = head;
        int idx = 1;

        while(temp.next != null){
            int pre = prev.val;
            int curr = temp.val;
            int next = temp.next.val;

            if((pre < curr && curr > next) || (pre > curr && curr < next))
                node.add(idx);

            idx++;
            prev = temp;
            temp = temp.next;
        }

        int n = node.size();

        if(n < 2)
            return  new int[]{-1 , -1};


        int max = node.get(n - 1) - node.get(0);
        
        int min = Integer.MAX_VALUE;

        for (int i = 1; i < n; i++) {
            min = Math.min(min, node.get(i) - node.get(i - 1));
        }


        return new int[]{min , max};
    }
}

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        temp = head.next
        prev = head

        node = []

        idx = 1

        while temp.next:
            curr = temp.val
            pre = prev.val
            upcoming = temp.next.val

            if (pre < curr and curr > upcoming) or (pre > curr and curr < upcoming) :
                node.append(idx)
            
            idx += 1
            prev = temp
            temp = temp.next

        if len(node) < 2 :
            return [-1 , -1]

        mx = node[len(node) - 1] - node[0]

        mn = math.inf

        for i in range(1 , len(node)):
            mn = min(mn , node[i] - node[i-1])

        return [mn , mx]
        